File: common/src/main/java/guepardoapps/lucahome/common/converter/JsonDataToTemperatureConverter.java
package guepardoapps.lucahome.common.converter;

import android.support.annotation.NonNull;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.UUID;

import guepardoapps.lucahome.common.classes.Temperature;
import guepardoapps.lucahome.common.utils.Logger;
import guepardoapps.lucahome.common.utils.StringHelper;

public final class JsonDataToTemperatureConverter implements IJsonDataConverter {
    private static final String Tag = JsonDataToTemperatureConverter.class.getSimpleName();
    private static final String SearchParameter = "{\"Temperature\":";

    private static final JsonDataToTemperatureConverter Singleton = new JsonDataToTemperatureConverter();

    public static JsonDataToTemperatureConverter getInstance() {
        return Singleton;
    }

    private JsonDataToTemperatureConverter() {
    }

    @Override
    public ArrayList<Temperature> GetList(@NonNull String[] jsonStringArray) {
        if (StringHelper.StringsAreEqual(jsonStringArray)) {
            return parseStringToList(jsonStringArray[0]);
        } else {
            String usedEntry = StringHelper.SelectString(jsonStringArray, SearchParameter);
            return parseStringToList(usedEntry);
        }
    }

    @Override
    public ArrayList<Temperature> GetList(@NonNull String jsonString) {
        return parseStringToList(jsonString);
    }

    private ArrayList<Temperature> parseStringToList(@NonNull String jsonString) {
        if (!jsonString.contains("Error")) {
            ArrayList<Temperature> list = new ArrayList<>();

            try {
                JSONObject jsonObject = new JSONObject(jsonString);
                JSONObject child = jsonObject.getJSONObject("Temperature");

                UUID uuid = UUID.fromString(child.getString("Uuid"));
                UUID roomUuid = UUID.fromString(child.getString("RoomUuid"));

                double temperatureValue = child.getDouble("Value");

                String sensorPath = child.getString("SensorPath");
                String graphPath = child.getString("GraphPath");

                Temperature temperature = new Temperature(uuid, roomUuid, temperatureValue, Calendar.getInstance(), sensorPath, Temperature.TemperatureType.Raspberry, graphPath);
                list.add(temperature);

            } catch (JSONException jsonException) {
                Logger.getInstance().Error(Tag, jsonException.getMessage());
            }
            return list;
        }

        Logger.getInstance().Error(Tag, jsonString + " has an error!");
        return new ArrayList<>();
    }
}

File: common/src/main/java/guepardoapps/lucahome/common/converter/temperature/JsonDataToTemperatureConverter.kt
package guepardoapps.lucahome.common.converter.temperature

import guepardoapps.lucahome.common.annotations.JsonKey
import guepardoapps.lucahome.common.enums.temperature.TemperatureType
import guepardoapps.lucahome.common.extensions.temperature.getJsonKey
import guepardoapps.lucahome.common.extensions.temperature.getPropertyJsonKey
import guepardoapps.lucahome.common.models.temperature.Temperature
import guepardoapps.lucahome.common.utils.Logger
import org.json.JSONArray
import org.json.JSONObject
import java.util.*

class JsonDataToTemperatureConverter {
    private val tag = JsonDataToTemperatureConverter::class.java.simpleName

    fun parse(jsonResponse: String): ArrayList<Temperature> {
        val list: ArrayList<Temperature> = ArrayList()

        if (jsonResponse.contains("Error")) {
            Logger.instance.error(tag, "Found error parsing: $jsonResponse")
            return list
        }

        try {
            val jsonObject = JSONObject(jsonResponse)

            val temperature = Temperature()
            val jsonKey: JsonKey = temperature.getJsonKey()
            val dataArray: JSONArray = jsonObject.getJSONArray(jsonKey.parent)

            for (index: Int in 0..dataArray.length()) {
                val value: JSONObject = dataArray.getJSONObject(index)
                val data: JSONObject = value.getJSONObject(Temperature::class.java.simpleName)

                val uuidJsonKey = temperature.getPropertyJsonKey(temperature::uuid.name)
                val valueJsonKey = temperature.getPropertyJsonKey(temperature::value.name)
                val areaJsonKey = temperature.getPropertyJsonKey(temperature::area.name)
                val sensorPathJsonKey = temperature.getPropertyJsonKey(temperature::sensorPath.name)
                val graphPathJsonKey = temperature.getPropertyJsonKey(temperature::graphPath.name)

                temperature.uuid = UUID.fromString(data.getString(uuidJsonKey.key))
                temperature.value = data.getDouble(valueJsonKey.key)
                temperature.area = data.getString(areaJsonKey.key)
                temperature.sensorPath = data.getString(sensorPathJsonKey.key)
                temperature.graphPath = data.getString(graphPathJsonKey.key)

                temperature.type = TemperatureType.Raspberry
                temperature.dateTime = Calendar.getInstance()

                list.add(temperature)
            }
        } catch (exception: Exception) {
            Logger.instance.error(tag, exception)
        } finally {
            return list
        }
    }
}